### PR TITLE
Document missing sync APIs on Deno Deploy

### PR DESCRIPTION
**Pull request by KnorpelSenf**
As reported in https://t.me/grammyjs/81279

### CODE REVIEW

**Review by xditya — APPROVED** [review]

**Review by roj1512 — APPROVED** [review]
LGTM. Thank you so much!

**Comment by KnorpelSenf** [comment]
How are you guys faster getting two reviews done before even CI completes … insane

File: site/docs/plugins/i18n.md [hunk 1]
```diff
@@ -656,6 +656,9 @@ await i18n.loadLocalesDir("locales"); // async version
 i18n.loadLocalesDirSync("locales-2"); // sync version
 ```
 
+> Note that certain environments require you to use the `async` version.
+> For example, Deno Deploy does not support synchronous file operations.
```
Comment (ngdangtu-vn) [inline site/docs/plugins/i18n.md:660]: Should we add the reference link for other developers (https://deno.com/deploy/docs/runtime-fs)

Comment (KnorpelSenf) [reply site/docs/plugins/i18n.md:671]: Can you perform any further changes in a subsequent PQ? As soon as something is ready for translation, we try not to alter the PQs anymore so that they can be translated more easily.

**Comment by roj1512** [comment]
@KnorpelSenf It's ghb!

**Comment by roj1512** [comment]
@ngdangtu-vn Please `deno fmt`.

File: site/docs/plugins/i18n.md (hunk 1 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:672]: PQ as **reason**?

I just thought that giving a reference will make other developers understand it easily.

**Comment by CikiMomogi** [comment]
@ngdangtu-vn Let me know if it's final :wink:

File: site/docs/plugins/i18n.md [hunk 2]
```diff
@@ -232,9 +232,14 @@ const bot = new Bot<MyContext>(""); // <-- put your bot token here (https://t.me
 // Continue reading to find out how to configure the instance.
 const i18n = new I18n<MyContext>({
   defaultLocale: "en", // see below for more information
+
+  // DOES NOT work in Deno Deploy
   directory: "locales", // Load all translation files from locales/.
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:237]:
```suggestion
  // Load all translation files from locales/. (Not working in Deno Deploy.)
  directory: "locales", 
```

File: site/docs/plugins/i18n.md [hunk 3]
```diff
@@ -232,9 +232,14 @@ const bot = new Bot<MyContext>(""); // <-- put your bot token here (https://t.me
...
 });
 
+// uncomment to asynchronously load translations in Deno Deploy
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:240]:
```suggestion
// Translation files loaded this way works on Deno Deploy, too.
```

File: site/docs/plugins/i18n.md [hunk 4]
```diff
@@ -391,13 +396,20 @@ interface SessionData {
...
   defaultLocale: "en",
   useSession: true, // whether to store user language in session
-  directory: "locales", // Load all translation files from locales/.
+
+  // DOES NOT work in Deno Deploy
+  directory: "locales",
 });
 
-const bot = new Bot<MyContext>(""); // <-- put your bot token here
+// uncomment to asynchronously load translations in Deno Deploy
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:411]:
```suggestion
// Translation files loaded this way works on Deno Deploy, too.
```

File: site/docs/plugins/i18n.md [hunk 5]
```diff
@@ -409,7 +421,7 @@ bot.use(
   }),
 );
 
-// Register the i18n middleware
+// Register the i18n middleware after session middleware
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:424]: This change is not related to the PR's goal

```suggestion
// Register the i18n middleware
```

File: site/docs/plugins/i18n.md [hunk 6]
```diff
@@ -391,13 +396,20 @@ interface SessionData {
 
 type MyContext = Context & SessionFlavor<SessionData> & I18nFlavor;
 
+const bot = new Bot<MyContext>(""); // <-- put your bot token here
+
+// Be awared, in Deno Deploy service,
+// we can only load fluent scripts after init I18n.
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:402]: Unnecessary imho. The user already has initialized a bot, and the two comments are extra.

```suggestion
```

File: site/docs/plugins/i18n.md (hunk 4 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:411]: I don't understand there is `too` in the end sentence. It's not like the `directory` method work with Deno Deploy right?

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:402]: The 2 comment lines are my fault, I supposed to remove it... But if you remove the line 399 the script will be broken if sb try to run it. The `bot` instance is never initialized until this line, isn't it?

File: site/docs/plugins/i18n.md (hunk 5 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:424]: I'm sorry, my bad.

File: site/docs/plugins/i18n.md (hunk 2 above)
Comment (roj1512) [reply site/docs/plugins/i18n.md:237]: Yes, it is.

File: site/docs/plugins/i18n.md (hunk 4 above)
Comment (roj1512) [reply site/docs/plugins/i18n.md:411]: The too means that it will work on Deno Deploy alongside other Deno environments.

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (roj1512) [reply site/docs/plugins/i18n.md:402]: And that piece of code's goal is not that. It's mentioned how to initialize the bot in the above snippets.

**Comment by ngdangtu-vn** [comment]
> @ngdangtu-vn Let me know if this PQ is final 😉

@CikiMomogi , I know it my fault for taking this longer than it should be. (dat face...)

File: site/docs/plugins/i18n.md (hunk 4 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:411]: I understand differently. I thought that there is other way that supposes to mention before that can work on Deno Deploy which we do not. Imho, the previous comment contains more meaning and clearer then this one. It gave a clear instruction on what to do and for what case.

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (roj1512) [reply site/docs/plugins/i18n.md:402]: I think they will understand if they RTFM from the beginning

**Comment by roj1512** [comment]
@ngdangtu-vn You can make us take less longer by directly accepting suggestions you are OK with.

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (ngdangtu-vn) [reply site/docs/plugins/i18n.md:402]: If you insist.

**Review by ngdangtu-vn — APPROVED** [review]

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (KnorpelSenf) [reply site/docs/plugins/i18n.md]: The existing example code snippets fall into two categories:
1) Short snippets. Illustrate a single thing only. One feature, one API call, one way of combining code, etc. These do not include imports and setup code.
2) Long snippets. Illustrate a setup, a complete bot, a minimal copy-paste-run example, etc. These do include imports and setup code.

Comment (roj1512) [reply site/docs/plugins/i18n.md]: Thanks for clarifying it the formal way, @KnorpelSenf!

File: site/docs/plugins/i18n.md [hunk 7]
```diff
@@ -232,9 +232,13 @@ const bot = new Bot<MyContext>(""); // <-- put your bot token here (https://t.me
 // Continue reading to find out how to configure the instance.
 const i18n = new I18n<MyContext>({
   defaultLocale: "en", // see below for more information
-  directory: "locales", // Load all translation files from locales/.
+  // Load all translation files from locales/. (Not working in Deno Deploy.)
+  directory: "locales",
 });
 
+// Translation files loaded this way works on Deno Deploy, too.
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:239]:
```suggestion
// Translation files loaded this way works in Deno Deploy, too.
```

File: site/docs/plugins/i18n.md [hunk 8]
```diff
@@ -394,10 +398,13 @@ type MyContext = Context & SessionFlavor<SessionData> & I18nFlavor;
...
   defaultLocale: "en",
   useSession: true, // whether to store user language in session
-  directory: "locales", // Load all translation files from locales/.
+
+  // DOES NOT work in Deno Deploy
+  directory: "locales",
 });
 
-const bot = new Bot<MyContext>(""); // <-- put your bot token here
+// Translation files loaded this way works on Deno Deploy, too.
```
Comment (roj1512) [inline site/docs/plugins/i18n.md:406]:
```suggestion
// Translation files loaded this way works in Deno Deploy, too.
```

**Review by roj1512 — APPROVED** [review]
LGTM.

**Comment by CikiMomogi** [comment]
Hmm.. No github actions? That's unusual :thinking:

**Comment by KnorpelSenf** [comment]
> Hmm.. No github actions? That's unusual 🤔

What do you mean?

**Comment by CikiMomogi** [comment]
Oh.. Glad it's working now.

**Comment by roj1512** [comment]
ShouldYouPlameYourPatience

File: site/docs/id/plugins/i18n.md [hunk 9]
```diff
@@ -396,10 +400,13 @@ type MyContext = Context & SessionFlavor<SessionData> & I18nFlavor;
...
   useSession: true, // Menentukan data bahasa user disimpan di session atau tidak
-  directory: "locales", // Muat locale dari direktori `locales`.
+
+  // TIDAK BEKERJA di Deno Deploy
+  directory: "locales",
 });
 
-const bot = new Bot<MyContext>(""); // <-- taruh token bot kamu di sini.
+// Untuk Deno Deploy, pakai cara berikut untuk memuat file terjemahan.
+// await i18n.loadLocalesDir("locales");
```
Comment (habemuscode) [inline site/docs/id/plugins/i18n.md:410]: Missing here: `const bot = new Bot<MyContext>(""); // <-- put your bot token here` translated

Comment (habemuscode) [reply site/docs/id/plugins/i18n.md:410]: I suppose is: `const bot = new Bot<MyContext>(""); // <-- taruh token bot kamu di sini.`

Comment (habemuscode) [reply site/docs/id/plugins/i18n.md:410]: Fixed by me, verify before merge

**Review by habemuscode — APPROVED** [review]

File: site/docs/id/plugins/i18n.md (hunk 9 above)
Comment (CikiMomogi) [reply site/docs/id/plugins/i18n.md:410]: See this discussion https://github.com/grammyjs/website/pull/515#discussion_r981603054

File: site/docs/plugins/i18n.md (hunk 6 above)
Comment (habemuscode) [reply site/docs/plugins/i18n.md]: For me, the same opinion as @ngdangtu-vn. It's not a snippet, it's a full code that include import statements, init session interfaces, init session, a few commands... and we are hidden something that someone will copy this and will not work.

I removed from all languages but 100% disagree, isn't a snippet, is a code to copy-paste and try

Full code:
```
import {
  Bot,
  Context,
  session,
  SessionFlavor,
} from "https://deno.land/x/grammy/mod.ts";
import { I18n, I18nFlavor } from "https://deno.land/x/grammy_i18n/mod.ts";

interface SessionData {
  __language_code?: string;
}

type MyContext = Context & SessionFlavor<SessionData> & I18nFlavor;

const i18n = new I18n<MyContext>({
  defaultLocale: "en",
  useSession: true, // whether to store user language in session

  // DOES NOT work in Deno Deploy
  directory: "locales",
});

// Translation files loaded this way works in Deno Deploy, too.
// await i18n.loadLocalesDir("locales");

// Remember to register `session` middleware before
// registering middleware of the i18n instance.
bot.use(
  session({
    initial: () => {
      return {};
    },
  }),
);

// Register the i18n middleware
bot.use(i18n);

bot.command("start", async (ctx) => {
  await ctx.reply(ctx.t("greeting"));
});

bot.command("language", async (ctx) => {
  if (ctx.match === "") {
    return await ctx.reply(ctx.t("language.specify-a-locale"));
  }

  // `i18n.locales` contains all the locales that have been registered
  if (!i18n.locales.includes(ctx.match)) {
    return await ctx.reply(ctx.t("language.invalid-locale"));
  }

  // `ctx.i18n.getLocale` returns the locale currently using.
  if ((await ctx.i18n.getLocale()) === ctx.match) {
    return await ctx.reply(ctx.t("language.already-set"));
  }

  await ctx.i18n.setLocale(ctx.match);
  await ctx.reply(ctx.t("language.language-set"));
});
```

**Review by roj1512 — APPROVED** [review]
